ios/Capacitor/Capacitor/CAPInstanceConfiguration.swift

```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+extension InstanceConfiguration {
+    @objc public func getPluginConfigValue(_ pluginId: String, _ configKey: String) -> Any? {
```

Comment:
Since this is now defined, should we deprecate `CAPPlugin.getConfigValue()` ?

Reply:
This has always been defined on `CAPConfig`. I moved it here for compatibility as `CAPPlugin.getConfigValue()` is just a pass-through to this method. (having two methods to do the same thing is potentially confusing but that's the way it's always worked)

Ideally, I think it'd be better if the plugin held its own configuration (which could be set at registration), at which point both of these methods could be deprecated. But if we pursue that approach it would be a follow-on piece of work.

Reply:
> Ideally, I think it'd be better if the plugin held its own configuration 

"held" meaning what exactly? What API do you propose for getting the plugin's config?

Reply:
Presently, the plugin's config is parsed from the file and held as a sub dictionary by the bridge's config. Like so:
```
{
"plugins": {
  "SplashScreen": {
    "launchShowDuration": 3000,
    "backgroundColor": "#ffffffff"
  }
}}
```
Which means that the plugins call `getConfigValue("backgroundColor")`, which calls `bridge.getPluginConfigValue("SplashScreen", "backgroundColor")` (or they can call `getPluginConfigValue` directly) and the bridge looks for that key path in its store to get the value. It's unnecessarily complicated.

Instead, the bridge could match up the name of the plugin and set the configuration at registration. So the plugin would have its own ivar that looks like:
```
{
  "launchShowDuration": 3000,
  "backgroundColor": "#ffffffff"
 }
```
So rather than reaching up the bridge for each and every value, it would function more like `CAPPluginCall`s `options`. And that opens the possibility for the plugin's config to adopt the same typed accessors like `getString`, `getBool`, etc.

Reply:
@ikeith That's what I figured, because that's exactly what I would do. 😁

One thing I did in my plugins was to implement a convenience API that looks first in a call's options and then falls back to the config. So:

```swift
func getString(from call: CAPPluginCall, orConfig fallback: Bool = false) -> String?
```

While you're rewriting so much API, have you given any thought to making the Swift API more iOS idiomatic? For example, this is idiomatic for JavaScript or Java, because there are no labeled parameters in those languages:

```swift
func getString(_ call: CAPPluginCall) -> String?
```

But this is idiomatic for iOS:

```swift
func getString(from call: CAPPluginCall) -> String?
```

Reply:
Well, the usage of `get` in the name isn't idiomatic for Swift in the first place so that isn't a great example. But, yes, moving toward a more fluent, idiomatic API is a goal but not an absolute one. It's a gradual process (some of which was done in #3678) and has to be balanced against backwards compatibility, unnecessarily disruption, and expectations of developers less familiar with the native conventions.

Reply:
You can always add idiomatic API that calls through to the non-idiomatic API. Or maybe vice versa is better since you may want to eventually deprecate the non-idiomatic API.

I guess the most idiomatic API would be to define functions on the config object on the plugin. Something like this:

```swift
class CAPPluginConfig {
  private config: [String: Any?]

  func string(withKey key: String) -> String? {
    return config[key] as? String
  }

  func int(withKey key: String) -> Int {
    return config[key] as? Int
  }

  // and so on...
}

class CAPPlugin {
  public config: CAPPluginConfig

  func getConfigString(_ key: String) -> String? {
    return config.string(withKey: key)
  }
}

public class MyPlugin: CAPPlugin {
  @objc func doSomething() {
    if let foo = config.string(withKey: "foo") {
      // hooray
    }
  }
}
```

Having wrestled with (and read all of the source code of) the current config API, I'd be happy to work on adding something like the above.

Reply:
In the example of CAPPluginCall, the idiomatic way to access the parameters would be `call.options["something] as? String ?? "default"`. The various `get` methods exist _solely_ as a convenience for developers who don't want/expect Swift idioms. So any change to the plugin would follow a similar pattern. Making Swift-idiomatic wrapper methods isn't the point.

Reply:
You are right, good point. I sometimes forget just how good Swift is.